nucleus/deployment/common/src/main/java/org/glassfish/deployment/common/ModuleDescriptor.java

```diff
@@ -145,7 +147,13 @@ public T getDescriptor() {
      * @param contextRoot the contextRoot
      */
     public void setContextRoot(String contextRoot) {
-        this.contextRoot = contextRoot;
+        VariableResolver resolver = new VariableResolver() {
```

Comment:
It's better to use the existing functionality in [TranslatedConfigView.expandValue()](https://github.com/eclipse-ee4j/glassfish/blob/master/nucleus/admin/config-api/src/main/java/org/glassfish/config/support/TranslatedConfigView.java#L52). This method is used to expand values in domain.xml and supports references to system properties and password aliases. In the future, we can also add support to MicroProfile Config values (e.g. also env variables).

I suggest creating a new static method in `TranslatedConfigView`, e.g. `expandValueForDescriptor` and delegate to the `expandValue()` method. This is just in case we need to add additional behavior when dealing with deployment descriptors. E.g. disable it with a config property like Payara did in this PR because it broke some TCK tests: https://github.com/payara/Payara/pull/4132.

Reply:
Hi @OndroMih,
Thanks for the hints ! I'll take a look and come back with some similar implementation.